File: python/leetcode39.py
from math import comb
from typing import List


class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        def dfs(candidates, target, combinations, tmp, index):
            if target < 0:
                return
            if target == 0:
                combinations.append(tmp[:])
                return
            for i in range(index, len(candidates)):
                if target - candidates[i] < 0:
                    return
                tmp.append(candidates[i])
                dfs(candidates, target - candidates[i], combinations, tmp, i)
                tmp.pop()

        combinations = []
        tmp = []
        candidates = sorted(candidates)
        dfs(candidates, target, combinations, tmp, 0)
        return combinations

File: cpp/leetcode39.cpp
#include <vector>

using namespace std;


class Solution {
public:
	void combination(vector< vector<int> >& res, vector<int>& candidates, vector<int> tmp, int target, int m) {
		if (target < 0)
			return;
		if (target == 0) {
			res.push_back(tmp);
		}
		for (int i = m; i < candidates.size(); i++) {
			tmp.push_back(candidates[i]);
			if (target - candidates[i] < 0)
				return;
			combination(res, candidates, tmp, target - candidates[i], i);
			tmp.pop_back();
		}
	}

	vector< vector<int> > combinationSum(vector<int>& candidates, int target) {
		if (candidates.size() == 0) {
			vector< vector<int> > res;
			return res;
		}
		vector< vector<int> > res;
		vector<int> tmp;
		sort(candidates.begin(), candidates.end());
		combination(res, candidates, tmp, target, 0);
		return res;
	}
};


int main(){
    vector<int> cans = {2, 3, 6, 7};
    Solution().combinationSum(cans, 7);
    return 0;
}